src/platform_impl/linux/kms/input.rs

```diff
@@ -0,0 +1,716 @@
+                    &mut (),
+                );
+            }
+
+            xkb_keymap::XKB_KEY_Control_L | xkb_keymap::XKB_KEY_Control_R => {
+                match state {
+                    ElementState::Pressed => $self.modifiers |= ModifiersState::CTRL,
+                    ElementState::Released => $self.modifiers.remove(ModifiersState::CTRL),
+                }
+
+                $callback(
+                    Event::WindowEvent {
+                        window_id: winid(),
+                        event: WindowEvent::ModifiersChanged($self.modifiers),
+                    },
+                    &mut (),
+                );
+            }
+
+            xkb_keymap::XKB_KEY_Meta_L | xkb_keymap::XKB_KEY_Meta_R => {
+                match state {
+                    ElementState::Pressed => $self.modifiers |= ModifiersState::LOGO,
+                    ElementState::Released => $self.modifiers.remove(ModifiersState::LOGO),
+                }
+
+                $callback(
+                    Event::WindowEvent {
+                        window_id: winid(),
+                        event: WindowEvent::ModifiersChanged($self.modifiers),
+                    },
+                    &mut (),
+                );
+            }
+
+            xkb_keymap::XKB_KEY_Sys_Req | xkb_keymap::XKB_KEY_Print => {
+                if $self.modifiers.is_empty() {
+                    $callback(
+                        Event::WindowEvent {
+                            window_id: winid(),
+                            event: WindowEvent::CloseRequested,
+                        },
+                        &mut (),
+                    );
+                }
+            }
+            _ => {}
+        }
+    }};
+}
+
+impl EventSource for LibinputInputBackend {
+    type Event = Event<'static, ()>;
+    type Metadata = ();
+    type Ret = ();
+    type Error = calloop::Error;
+
+    fn process_events<F>(
+        &mut self,
+        _: Readiness,
+        token: Token,
+        mut callback: F,
+    ) -> calloop::Result<PostAction>
+    where
+        F: FnMut(Self::Event, &mut ()) -> Self::Ret,
+    {
+        if self.token.filter(|t| *t == token).is_some() {
+            self.context.dispatch()?;
+
+            for event in &mut self.context {
+                match event {
+                    input::Event::Device(ev) => handle_device_event!(ev, callback),
+                    input::Event::Touch(ev) => handle_touch_event!(self, ev, callback),
+                    input::Event::Tablet(ev) => handle_tablet_tool_event!(self, ev, callback),
+                    input::Event::Pointer(ev) => handle_pointer_event!(self, ev, callback),
+                    input::Event::Keyboard(ev) => handle_keyboard_event!(self, ev, callback),
+                    _ => {}
+                }
+            }
+        }
+        Ok(PostAction::Continue)
+    }
+
+    fn register(&mut self, poll: &mut Poll, factory: &mut TokenFactory) -> calloop::Result<()> {
```

Comment:
Maybe instead of using `poll` directly, we can wrap the `libinput` handling in the [`calloop::Generic`](https://docs.rs/calloop/latest/calloop/generic/struct.Generic.html) type and use that to handle things?